x-pack/legacy/plugins/uptime/server/lib/adapters/monitor_states/elasticsearch_monitor_states_adapter.ts

```diff
@@ -29,11 +29,119 @@ interface LegacyMonitorStatesQueryResult {
   }
 
+  // This query returns the most recent check groups for a given
+  // monitor ID.
+  private async runLegacyMonitorStatesRecentCheckGroupsQuery(
+    request: any,
+    query: any,
+    searchAfter?: any
+  ): Promise<LegacyMonitorStatesRecentCheckGroupsQueryResult> {
+    const body = {
+      query: {
+        bool: {
+          filter: [
+            {
+              // We check for summary.up to ensure that the check group
+              // is complete. Summary fields are only present on
+              // completed check groups.
+              exists: {
+                field: 'summary.up',
+              },
+            },
+            query,
+          ],
+        },
+      },
+      sort: [
+        {
+          '@timestamp': 'desc',
+        },
+      ],
+      size: 0,
+      aggs: {
+        monitors: {
+          composite: {
+            size: LEGACY_STATES_QUERY_SIZE,
+            sources: [
+              {
+                monitor_id: {
+                  terms: {
+                    field: 'monitor.id',
+                  },
+                },
+              },
+            ],
+          },
+          aggs: {
+            top: {
+              top_hits: {
+                sort: [
+                  {
+                    '@timestamp': 'desc',
+                  },
+                ],
+                _source: {
+                  includes: ['monitor.check_group', '@timestamp', 'agent.id'],
+                },
+                // The idea here is that we want to get enough documents to get all
+                // possible agent IDs. Doing that in a deterministic way is hard,
+                // but all agent IDs should be represented in the top 50 results in most cases.
+                // There's an edge case here where a user has accidentally configured
+                // two agents to run on different schedules, but that's an issue on the user side.
+                size: 50,
+              },
+            },
+          },
+        },
+      },
+    };
+
+    if (searchAfter) {
+      set(body, 'aggs.monitors.composite.after', searchAfter);
+    }
+
+    const params = {
+      index: INDEX_NAMES.HEARTBEAT,
+      body,
+    };
+
+    const result = await this.database.search(request, params);
+
+    const checkGroups = result.aggregations.monitors.buckets.flatMap((b: any) => {
+      const topHits = get<any[]>(b, 'top.hits.hits', []);
+
+      const latestAgentGroup: { [key: string]: { timestamp: string; checkGroup: string } } = {};
+      topHits.forEach(({ _source: source }) => {
+        // We set the agent group to the first thing we see since it's already sorted
+        // by timestamp descending
+        if (!latestAgentGroup[source.agent.id]) {
+          latestAgentGroup[source.agent.id] = {
+            timestamp: source['@timestamp'],
+            checkGroup: source.monitor.check_group,
+          };
+        }
+      });
+
+      return Object.values(latestAgentGroup).map(lag => lag.checkGroup);
```

Comment:
Another subjective point but it seems like `lag` has some ambiguity, at first glance a reader could think it's related to latency of some kind. The pattern here also typically consists of a collections named with plurality, like `monitorIds` and a singular iterator name like `monitorId`. In this case, we're only accessing one property that is defined in the typing, so why don't we just destructure it instead of declaring a new symbol?

```suggestion
      return Object.values(latestAgentGroup).map(({ checkGroup }) => checkGroup);
```